assets/src/models/createDetourMachine.ts

```diff
@@ -1,17 +1,139 @@
+          routeId: RouteId
+          routePatternId: undefined
+        }
+      | {
+          // Caller has no prior selection
+          routeId: undefined
+          routePatternId: undefined
+        }
+
+    events:
+      | { type: "detour.edit.done" }
+      | { type: "detour.edit.resume" }
+      | { type: "detour.route-pattern.open" }
+      | { type: "detour.route-pattern.done" }
+      | { type: "detour.route-pattern.select-route"; routeId: RouteId }
+      | {
+          type: "detour.route-pattern.select-pattern"
+          routePatternId: RoutePatternId
+        }
+  },
+  actors: {
+    "fetch-route-patterns": fromPromise<
+      Awaited<ReturnType<typeof fetchRoutePatterns>>,
+      { routeId?: RouteId }
+    >(async ({ input: { routeId } }) => {
+      if (routeId) {
+        return fetchRoutePatterns(routeId)
+      } else {
+        // Hmm, this could leave us in the lurch if we don't handle it,
+        // I'd like to make this impossible but I can't get typescript to infer
+        // the context types but it's not letting me
+        throw "No Route ID"
+      }
+    }),
+  },
+  actions: {
+    "set.route-pattern": assign({
+      routePatternId: (_, params: { routePatternId: RoutePatternId }) =>
+        params.routePatternId,
+    }),
+    "set.route-id": assign({
+      routeId: (_, params: { routeId: RouteId }) => params.routeId,
+    }),
   },
 }).createMachine({
   id: "Detours Machine",
+  context: ({ input }) => ({
+    ...input,
+  }),
 
   initial: "Detour Drawing",
   states: {
     "Detour Drawing": {
       initial: "Editing",
       states: {
+        "Pick Route Pattern": {
+          initial: "Pick Route ID",
+          on: {
+            "detour.route-pattern.select-route": {
+              target: ".Load Route Patterns",
+              actions: assign({
+                routeId: ({ event }) => event.routeId,
+                routePatternId: undefined,
+              }),
+            },
+          },
+          states: {
+            "Pick Route ID": {
+              /**
+               * This is intentionally left empty because the only way to
+               * "Pick Route Pattern" is to pick a Route ID, but the Route ID
+               * is allowed to change at any point while inside the
+               * "#Detours Machine.Detour Drawing.Pick Route Pattern" state.
+               */
+            },
+            "Load Route Patterns": {
+              invoke: {
+                src: "fetch-route-patterns",
+
+                input: ({ context: { routeId } }) => ({ routeId: routeId }),
+                onDone: {
+                  target: "Finalize Route Pattern",
+                  actions: assign({
+                    routePatterns: ({ event }) => event.output,
+                    routePatternId: ({ event }) =>
+                      event.output.find((pattern) => pattern.directionId === 1)
+                        ?.id,
+                  }),
+                },
+              },
+            },
```

Comment:
Checking my comprehension here: it seems that this state is a transitional state that doesn't wait for any inputs before moving along? Like, "Load Route Patterns" whenever the "select-route" event happens, and then it invokes the "fetch-route-patterns" logic. How does the onDone get triggered for this state? Does the state automatically go into "Finalize Route Pattern" when the fetch finishes? If that's the case, should "Load Route Patterns" be a state if it doesn't wait for an event / input?

Reply:
Relevant docs: https://stately.ai/docs/invoke#ondone

---

I could theoretically see this state & actor being merged into `Pick Route ID`, where this actor is invoked every time that we enter that state, and we only transition out of that state once we have route patterns loaded.

Currently, I _think_ I prefer this as it's own state because I like that what the machine is doing is explicitly described, but I'm open to chatting about this more to figure out if that's desired!

I think if we did what you're saying, it would look like this?

<details>
<summary>Code</summary>

```ts
"Pick Route Pattern": {
  initial: "Pick Route ID",
  on: {
    "detour.route-pattern.select-route": {
      target: ".Pick Route ID",
      actions: assign({
        routeId: ({ event }) => event.routeId,
        routePatternId: undefined,
      }),
    },
  },
  states: {
    "Pick Route ID": {
      invoke: {
        src: "fetch-route-patterns",

        input: ({ context: { routeId } }) => ({ routeId: routeId }),
        onDone: {
          target: "Finalize Route Pattern",
          actions: assign({
            routePatterns: ({ event }) => event.output,
            routePatternId: ({ event }) =>
              event.output.find((pattern) => pattern.directionId === 1)
                ?.id,
          }),
        },
      },
    },
    "Finalize Route Pattern": {
      on: {
        "detour.route-pattern.done": {
          guard: ({ context }) => context.routePatternId !== undefined,
          target: "Done",
        },
        "detour.route-pattern.select-pattern": {
          actions: {
            type: "set.route-pattern",
            params: ({ event: { routePatternId } }) => ({
              routePatternId,
            }),
          },
        },
      },
    },
    Done: {
      type: "final",
    },
  },
  onDone: {
    target: "Editing",
  },
}
```

</details>

I'm still a little on the fence, but I somewhat like the above example code, so lets chat about it! Do you think that the implicit state in the above example reads better or anything else?

Reply:
I'm warming up to this more and more FYI!
But I'm still feeling like there is _some_ reason to keep the explicit state that I can't put my finger on (and my intuition could be wrong here even if I do remember it)

Reply:
One issue here, it makes the states read a little bit stranger when adding nested loading states, see
```ts
snapshot.matches({
  "Detour Drawing": {
    "Pick Route Pattern": { "Pick Route ID": "Loading" },
  },
})
```
vs
```ts
snapshot.matches({
  "Detour Drawing": {
    "Pick Route Pattern": { "Load Route Patterns": "Loading" },
  },
})
```

Reply:
This makes it seem like the Route ID is loading, when it's the patterns that are loading.